### ISSUE TITLE
FRAGMOS - TermsChange shall include Legal Agreement

### ISSUE THREAD

**Issue by JBZ-Fragmos** [finos/common-domain-model#2100]
**PROBLEM STATEMENT**
As of today it is impossible to amend other Trades items than “Product” and “Ancillary Parties” 
– whereas a whole Trade agreement may also contain other items notably the ones related to “Legal Agreement” (pass through Trade via “contractFormation” Instruction) 

![image](https://user-images.githubusercontent.com/92978493/232550653-31f4f2d3-0e47-4f36-bd54-680c6a75a906.png)

**Use Case** : 
- in legacy world (out of CDM) say you have signed by error a Long Form Confirmation, so you wish to send another version that will Cancel & Replace previous one if the other Party would agree to sign it 
- this is precisely such common use case that what we cannot replicate via CDM today, because there is no way to amend the items attached to a Trade via legalAgreement root... 

**PROPOSAL**
we need to add “Legal Agreement” as an attribute of “Term Change” Instruction, for **consistency** purposes so that such instruction will not only permit to amend core economics via product root, but also contractDetails in trade (via legalAgreeme,nt root) - if not, as of today, one cannot represent an amendment of the Legal Agreement attached to a given trade, say for instance AdditionalTerms need to be amended : today you cannot... to do so, we need being able to use the termsChange Instruction on larger scope than today i.e. that could encompass as well LegalAgreement"

![image](https://user-images.githubusercontent.com/92978493/232550902-31c42386-b314-4cef-97ab-b34b5a3e5a3a.png)

**Comment by tomhealey-icma** [comment]
Strangely, TermsChangeInstruction has only three optional parameters with a condition that AtLeastOneOf: (product exists or ancillaryParty exists or adjustment exists). Product should be required and at least one of ancillaryParty exists or adjustment exists. That aside, the intention is on modifying the tradable product. To do what you reference, especially since you indicated it's a cancel and replace, you would be Create_ContractFormationInstruction( item -> trade -> contractDetails -> documentation ) to change the legal agreement together with the appropriate workflow.

**Comment by JBZ-Fragmos** [comment]
let me rephrase the business need

as of today

- you can build an agreement on **trade Core Economics**
- **builder** object is the **"execution" instruction**, that will result in a Trade with mostly "Product" path values being populated
- now, if you would like **to amend** such original agreement - **you can do it**, precisely by using the **"termsChange"** instruction that will permit you to cancel/replace items in **Product**, thus to amend the Trade accordingly

now please consider the same (i.e. how to build an agreement then how to amend it when need be)

- as regards the agreement on **trade Documentation**, that is actually of type "LegalAgreement", when attached as aTrade attribute, under "contractDetail" (below) 
- builder is the object you mention i.e. **"contractFormation" instruction** that will result in "contractDetails->LegalAgreement" path values being populated in the Trade
- now, if you would like **to amend** such original agreement - **you cannot do it**, and here is the BUSINESS NEED : you cannot because **"termsChange"** instruction does not permit you to cancel/replace items in **LegalAgreement**

hence my update proposal, that is for consistency reason... to include LegalAgreement as well in the scope of items that termsChange will permit to amend as part of subsequent workflow after a Trade has been built... it is only missing piece of the puzzle...

if to recap this in few words, would say 
"trade = Product + Documentation", 
therefore there is no reason that "termsChange" which functional usage is to amend trade items, would be "termsChange= Product only"... for symetrical reasons, shoudl be "termsChange = Product + Documentation", 

Else how would you do if you want to amend Documentation path values in your Trade ?

![image](https://user-images.githubusercontent.com/92978493/232703936-6f229c3c-d7b0-47a1-babf-7db4400661e9.png)

**Comment by dshoneisda** [comment]
I support the concept in terms of the business use case of amending confirmation elections as I can see that there will be times that traders will agree to override certain default house elections on specific transactions. However, I think we need to be careful around allowing all of the attributes to be amended. For example I don't think that it should be possible to amend the type of legal agreement governing the trade i.e. from one type of master to another, as that has many potential real-world implications throughout the lifecycle including reporting and should enforce a cancellation and new booking.
So I think any update to allow for amendment in this area should be qualified to only allow those elements that make sense

**Comment by JBZ-Fragmos** [comment]
thanks for reviewing 

I fully agree with your comment actually in my mind I intended to stay at trade level i.e. agree with you the purpose shall not be to amend an agreement that is higher in hierarchy compared to the trade e.g. MA, MCA, etc.

taking into account your good feedback, here is **refactored proposal**

in order to ensure exact scope you can amend is the scope of the trade and no more (that is all in all what we want here i.e.to beable to amend both product terms and additional terms of the transaction) the new attribute shall **not** be **legalAgreement**, but instead shoudl be further down the path attached to this node type i.e. **transactionAdditionalTerms**, which name precisely indicate exactly what we are targeting i.e. **trade=transaction level**

![image](https://github.com/finos/common-domain-model/assets/92978493/ffdbf4af-ffb4-488c-83bd-9131c6771e83)

looks really good to me now as proposed : what do you think now ? good for you as well ?

**Comment by dshoneisda** [comment]
Seems reasonable to me
any additional thoughts on this? @tomhealey-icma @chrisisla @rvincentISDA @eacunaISDA @lolabeis

**Comment by tomhealey-icma** [comment]
The problem is that the legalAgreement is attached to the contractDetails at the trade level which is above the product. I've raised this issue before, a product, in many cases is defined by the legalAgreement, so I've proposed adding contractDetails to product or as it's been suggested, create a layer above product and below trade such that product + contractDetails = LegalProduct. Before adding legalAgreement to TermsChange I believe we need to address this problem.

**Comment by dshoneisda** [comment]
@Oblongs @lolabeis will the proposals for the asset and observable and product refactorings address Tom's suggestion and lead to resolution for JB?

**Comment by chrisisla** [comment]
Hi @JBZ-Fragmos ,
Apologies for the delay in reply, to be blunt there are too many active Issues at the moment so it is hard to keep track of where we are...
In relation to your second question, I think the opposite of what you are suggesting is true i.e. the default should be to NOT just "assume" something is accepted if one maintainer has reviewed and approved a proposal. Admittedly this is not ideal but we can't say consensus has been reached if only one person has looked at it.
Thanks.

**Comment by JBZ-Fragmos** [comment]
will no more insist, but this is only consistent consideration : 
there is a new item which may be executed as part of a trade agreement that is transactionAdditionalTerms (= Product Model)
consequently, there is a need to represent as well how such agreement could be updated, that is having transactionAdditionalTerms in tremsChange (= Event Model)

in my opinion, having one without the other one, is nothing else than having a "hole" in the modelling... also preventing CDM to represent reality : parties may cancel/replace prior execution of agreements

the fact other components might require similar modelling and might be missing today shall not block the one already defined and identified to be properly released

**Comment by tomhealey-icma** [comment]
There was a rather strong objection raised when I proposed adding contractDetails to product which resulted in an alternate proposal as I mentioned above. Regarding,  transactionAdditionalTerms, from what I see it only exists in the legal model not in the product model so as I said, the termsChangeInstruction only accepts product input and not a trade which is where the legalAgreement would reside.

**Comment by JBZ-Fragmos** [comment]
sorry this is not the point, which eventually is per se a question of being logic in terms of modelling i.e. ensure Event Model is consistent per se, I mean please forget any business meaning of objects

- just consider it as below, in _abstract manner_ 
   - GIVEN "**A**" exist as an attribute of "**X**" as result of an **execution** or **contractFormation** Event
   - THEN logically we need for consistency purposes that **termsChange** Event must permit as need be to **update "A" i**n **"X"**

- now if we may replace in the above description :
   - "A" by "**transactionAdditionalTerms**"
   - "X" by "**Trade->ContractDetails->LegalAgreement->[...]->transactionAdditionalTerms**"
 
hope it does clarify the need, because I must admit I have no more ideas how to further clarify it, and it is clear this is not at all contradictory to another need, being to get similar feature elsewhere as well...

**Comment by rvincentISDA** [comment]
Per the length of the latest feedback I recommend to take this conversation to a CDM working group with coverage across associations. The Derivative Products and Business Events working group is a prime candidate. 

The termChangeInstructions was introduced a long time ago as a temporary catch-all whilst CDM continued growing. It requires further considerations at least across 4 data domains: 
i.	The definition of the product
ii.	The execution and performance of the trade
iii.	The management of the trade 
iv.	The enablement of the trade. 

Arguably _termchange_ could be restricted from touching item iv) on its own as it could compromise i),  ii) or iii). 

TransactionAdditionalTerms pertains to item iii) hence the reason why it was moved out of domain  i). In Efect _TradableProduct_ is slowly converging toward items i) and ii). There is still plenty to inspect. 

Similarly, analysis and further modelling work would be welcome to disentangle contractDetails and rationalise the overlap across tradableProduct e.g. counterparty vs. contractualParty  or Tom's point on aspects of contractDetails necessary to define what is being trade (e.g. "LegalProduc")

_TransactionAdditionalTerms_  predicates aspect of the management of the trade and it should be possible to amend them. However, the proposed PR is insufficient because it lacks input data to deal with the unbounded cardinality of the _documentation_ attribute under _contractDetails_. If you have multiple _LegalAgreements_ attached to the trade, how do you know which one is impacted by the new _transactionAdditionalTerms_. In addition should there not be other code change to reflect the amendment.  I failed to found them .

**Comment by JBZ-Fragmos** [comment]
i agree the topic is part of potentially larger scope than the one covered by the PR, that is when taking into account above feedback from Ruddy also previous ones from Tom to some extent 

yet the specific scope of this PR is clear and consistent per se : 
purpose is to permit to amend a particular object which may require to be amened when need be

this is why I still consider the particular PR already makes sense, is valid stand-alone, thus shall be released, that _withtout blocking_ any larger considerations

all in all, could we agree to say that larger considerations shall not block particular PR when it is established such PR is part of the larger set at stake and is moving in the good direction without undermining generic design of the larger scope to which it belongs

in that regards, forgive me if repeating previous rationale but current PR intends to be logical consequence of the recent release of TransactionAdditionalTerms : since this object exists as part of a Trade in path Trade->ContractDetails->LegalAgreement->[...]->transactionAdditionalTerms then we logically need to update it where it is - hence the proposal.

one may also step back and consider that today, **Trade** contains these **2 types of info** : 
- **core economics**, say it is modelled by **Product** 
- AND **legal additional terms** - say it is modelled by **TransactionAdditionalTerms**

then refactoring of termsChange is only the _**logical mirroring**_ of such Trade structure inside Event Model for the purpose of amending it, nothing that may block anything else, eventually :
- **Product** is already present in termsChange in order to amend core economics
- SIMILARLY **TransactionAdditionalTerms** is currently missing for the same for **legal additional terms**

**Comment by lolabeis** [comment]
@JBZ-Fragmos @tomhealey-icma @dshoneisda

First, allow me to clarify the current situation re `transactionAdditionalTerms`:
- Following the refactoring of Extraordinary Events, `transactionAdditionalTerms` is now part of the `Agreement` model (in the `legalagreement` namespace) but **no longer** part of the `Product` model  (in the `product` namespace). Stricto sensu, this last statement means that no path starting from `Product` would include any `transactionAdditionalTerms`.
- `transactionAdditionalTerms` **is** included in the `Trade` model: i.e. there are paths starting from `Trade` that include  `transactionAdditionalTerms` - but note that `Trade` is not part of the `Product` model, it sits above (in the `event` namespace).

Now, that doesn't imply that a user shouldn't be able to modify elements of a `Trade` that aren't in `Product` or  its underlying attributes - for instance, `PartyChangePrimitive` does that with the `Party` attributes. In the end, if all attributes of `Trade` are subject to change under a lifecycle event, there should be a primitive event that modifies it. But the contention is whether `TermsChangePrimitive` is the right one for it.

There are 3 possible approaches:
1. Include it in `TermsChangePrimitive` (as currently proposed). Pro: Uses an existing primitive, and is consistent with that primitive's purpose (to enable the amendment of certain trade terms). Con: A single primitive that touches attributes that are now positioned in very distinct parts of the model.
2. Develop a new primitive designed to update only `transactionAdditionalTerms`. Pro: Cleaner primitive design, separation aligned onto where attributes are positioned. Con: Needs a new primitive that may not have any use other than in the context of an amendment. Also overlap between this new primitive and `ContractFormationPrimitive` (in theory, both could modify `transactionAdditionalTerms`).
3. Repurpose `ContractFormationPrimitive` to accommodate that use case. Pro: Uses an existing primitive that already touches the same part of the model. Con: That primitive's purpose becomes quite broad, and the qualification logic to identify an amendment would become more complex (it would need to assert against that broad-based primitive).

These are just resolution directions - there may be a hybrid approach that keeps the Pros and mitigates the Cons. But I propose to focus the discussion on that.

The "larger" issue that is already being discussed in #2216 still needs to be resolved, but separately from the present one which proposes a reasonable change based on the model as currently designed.

**Comment by JBZ-Fragmos** [comment]
FYI, taking into account last comment from Ruddy, i've also desifned "v2" below - to be discussed as well 

![image](https://github.com/finos/common-domain-model/assets/92978493/d8c3ed18-7079-4b50-b5f1-e922bd97dff9)

**Comment by chrisisla** [comment]
My preference would be option 2. 

Would there be any issue with the new primitive being able to update anything under ```trade -> contractDetails``` and not just being specific for ```transactionAdditionalTerms```? i.e. create something more along the lines of a ```ContractAmendmnentPrimiitive```?

**Comment by JBZ-Fragmos** [comment]
i will draft a proposal aligned with option2, then we will review it as new starting point for our discussions

**Comment by lolabeis** [comment]
@JBZ-Fragmos 
> i will draft a new proposal aligned with option2

Thinking further about this, I think the right design sits somewhere between options 2 and 3. This is because of the _Con_ in option 2 that any new primitive would overlap with the existing `ContractFormationPrimitive`.

Specifically, option 3 should be the starting point, but then recognising that an all-or-nothing update to the `LegalAgreement` may be too blunt an instrument, we could split that primitive into smaller ones that update different parts of that attribute?

@chrisisla 
> Would there be any issue with the new primitive being able to update anything under trade -> contractDetails and not just being specific for transactionAdditionalTerms?

It seems your requirement fits better with option 3 then, hence my point above.

**Comment by rvincentISDA** [comment]
My understanding is that contract_Formation is the primitive that marks the border between pre and post-trade. Repurposing it will be too blunt and could entails months of work as it condition many use cases, for example EMIR vs. MIFID. A targeted tactical option 2 is preferable and is likely to the seed the demarcation across the data domains I suggested above.

**Comment by JBZ-Fragmos** [comment]
all in all, my understanding is that the consensus is closer to "**option n°2**" above described... 

accordingly, here is a new proposal, 
- please let me know if answering most points raised @lolabeis @tomhealey-icma @rvincentISDA @chrisisla 

will finalize related new PR once we agree on items as described below.

**REQUIREMENTS**

requirements above discussed mainly consist in ensure such Instruction only permits :

- to update **Legal Agreement** terms **restricted** to the particular terms at stake at **Trade level**
   - that is to say the **root object** to update as part of the Instruction shall be **"trade->contractDetail->documentation"** ; because "documentation" in such path is precisely the name of Legal Agreement when being used as a Trade attribute.

- to update **Related Agreements** terms **also restricted** to the particular terms at stake at **Trade level** 
   - which means only the terms used for the purpose of **determining which Related Agreement is applicable** to the Trade, while **exclusing** to update **any content of such Related Agreement**
      - _use case / example_ : it shall be permitted to update the effectiveDate of Related Agreement such as "Master Agreement" because this is typical information used in Trade Conformation for the purpose of referencing which Master Agreement is applicable to the particular Trade, but it cannot modify the content of such Master Agreement)

**graphical view of scope at stake, as per requirements**

- the Instruction shall permit to update any terms from root below,
- exlcuding_ the components we have framed


**PROPOSAL**

![image](https://github.com/finos/common-domain-model/assets/92978493/bedb9f7f-dda3-442c-b328-0d94afb980ee)

**Textual view of the new Type** 

- only attribute is of type **LegalAgreement**
- the **two conditions** for the purpose of specifying exact scope of update is aligned with requirements not to permit any update of a documentation content beyond Trade level

![image](https://github.com/finos/common-domain-model/assets/92978493/f44345a7-e067-4479-9403-80e29a028d63)

**consequently, this new type shall be a new attribute of Primitive Instruction**
![image](https://github.com/finos/common-domain-model/assets/92978493/ba4dbeff-4522-4e30-a208-24be3f3a64d0)

**graphical view of the proposal**

we may see that scope of update is aligned with original scope agreed that might be at target if there is a need to update the original agreement, that with specific conditions we cannot see on graphics which restrict such scope to ensure detailed terms updated remain at Trade level
![image](https://github.com/finos/common-domain-model/assets/92978493/21619dfc-8f60-44e0-89f1-4adccac8ff37)

**Comment by JBZ-Fragmos** [comment]
not needed as priority at this stage
